nifi-registry-core/nifi-registry-jetty/src/main/java/org/apache/nifi/registry/jetty/JettyServer.java

```diff
@@ -106,6 +112,37 @@ public JettyServer(final NiFiRegistryProperties properties, final CryptoKeyProvi
         }
     }
 
+    /**
+     * Returns a File object for the directory containing NIFI documentation.
+     * <p>
+     * Formerly, if the docsDirectory did not exist NIFI would fail to start
+     * with an IllegalStateException and a rather unhelpful log message.
+     * NIFI-2184 updates the process such that if the docsDirectory does not
+     * exist an attempt will be made to create the directory. If that is
+     * successful NIFI will no longer fail and will start successfully barring
+     * any other errors. The side effect of the docsDirectory not being present
+     * is that the documentation links under the 'General' portion of the help
+     * page will not be accessible, but at least the process will be running.
+     *
+     * @param docsDirectory Name of documentation directory in installation directory.
+     * @return A File object to the documentation directory; else startUpFailure called.
+     */
+    private File getDocsDir(final String docsDirectory) {
```

Comment:
I'm wondering if this change was really needed... looking at lines 383-394 of the original code, it looks like the logic there was basically the same that is in this method.

Was there anything else that was being fixed by changing the way the docs webapp was being loaded?

Reply:
The main thing here to change the way the docs are loaded was by using a ServletHolder instead of a ResourceHandler. The getDocsDir method was extracted in NiFi to this separate method, so I basically just reused that.

Reply:
I guess my question was... did we need to change the way the docs webapp was loaded in order for something with the security headers to work correctly? 

I'm only asking since I don't think there was any existing issue with the docs webapp.

Reply:
Yes, without using a ServletHolder for the docs webapp, the Jetty filters and thereby the security headers are not applied correctly. This is something I experienced when attempting the change in NiFi. Basically when using a ResourceHandler, when the request is made to the docs the filters are not triggered. I did research into this, possibly emailed the Jetty dev distro, and eventually found that I had to use a ServletHolder.